src/sse42_utf8check.rs

```diff
@@ -0,0 +1,251 @@
+use crate::*;
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
+/*
+ * legal utf-8 byte sequence
+ * http://www.unicode.org/versions/Unicode6.0.0/ch03.pdf - page 94
+ *
+ *  Code Points        1st       2s       3s       4s
+ * U+0000..U+007F     00..7F
+ * U+0080..U+07FF     C2..DF   80..BF
+ * U+0800..U+0FFF     E0       A0..BF   80..BF
+ * U+1000..U+CFFF     E1..EC   80..BF   80..BF
+ * U+D000..U+D7FF     ED       80..9F   80..BF
+ * U+E000..U+FFFF     EE..EF   80..BF   80..BF
+ * U+10000..U+3FFFF   F0       90..BF   80..BF   80..BF
+ * U+40000..U+FFFFF   F1..F3   80..BF   80..BF   80..BF
+ * U+100000..U+10FFFF F4       80..8F   80..BF   80..BF
+ *
+ */
+
+// all byte values must be no larger than 0xF4
+
+/*****************************/
+#[cfg_attr(not(feature = "no-inline"), inline)]
+fn push_last_byte_of_a_to_b(a: __m128i, b: __m128i) -> __m128i {
+    unsafe { _mm_alignr_epi8(b, a, 15) }
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline)]
+fn push_last_2bytes_of_a_to_b(a: __m128i, b: __m128i) -> __m128i {
+    unsafe { _mm_alignr_epi8(b, a, 14) }
+}
+
+// all byte values must be no larger than 0xF4
+#[cfg_attr(not(feature = "no-inline"), inline)]
+fn avxcheck_smaller_than_0xf4(current_bytes: __m128i, has_error: &mut __m128i) {
```

Comment:
the avx* names are a misnomer, I held off on a renaming party (for now) thinking we'd need to change both places (AVX2 and SSE42)

Reply:
Ja with rust we have modules so we don't really need function prefixes :) good catch!